### PR TITLE
Improve error message for missing mzbuild services

### LINKED ISSUE #5020: mzconduct failure message when attempting to build new services is confusing
`mzconduct` / `mzcompose` ignore files that git doesn't know about and then print a very confusing
error message when they cannot find a service specified by `mzbuild`. A service which they are
ignoring.

When adding new services, my typical workflow is to create the directory for the service, add an
`mzbuild.yml`, `Dockerfile`, source code and then test create the image by running: `docker build
-f Dockerfile .`.

If the image builds cleanly, I then go and add this service to my `mzcompose.yml` file. For
example:

```
services:
+  create-views:
+    mzbuild: wikirecent-create-views
```

When I go to run `mzconduct`, I get the following error message:

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/chris/materialize/materialize/misc/python/materialize/cli/mzcompose.py", line 200, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/home/chris/materialize/materialize/misc/python/materialize/cli/mzcompose.py", line 68, in main
    composes.append(build_compose_file(repo, args.command, config_file))
  File "/home/chris/materialize/materialize/misc/python/materialize/cli/mzcompose.py", line 124, in build_compose_file
    say(d.spec())
  File "/home/chris/materialize/materialize/misc/python/materialize/mzbuild.py", line 426, in spec
    return self.image.docker_name(f"mzbuild-{self.fingerprint()}")
  File "/home/chris/materialize/materialize/misc/python/materialize/mzbuild.py", line 565, in fingerprint
    with open(abs_path, "rb") as f:
IsADirectoryError: [Errno 21] Is a directory: '/home/chris/materialize/materialize/test/sqllogictest/sqlite'
```

One solution is to run `git add` on the service that I just created. It would be nice, however, if
this error message were a little more helpful.

### PR DESCRIPTION
**Pull request by cirego**
When adding new services, it's common to forget to `git add` the
directory that contains the mzbuild.yml file. Before this change, that
would result in an error message about `sqllogictest/sqlite` being a
directory. After this change, the user should be made aware that the
problem is between their mzcompose.yml, mzbuild.yml and git.

The proper fix for this is to run `git add` on the mzbuild.yml file.

Fixes #5020

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5269)

### CODE REVIEW

**Comment by benesch** [comment]
Thanks for digging into this so thoroughly! This looks like a bona fide mzbuild bug though. The intent was never that you'd have to run `git add` to get things going. Let me PR a fix at you shortly.

**Comment by cirego** [comment]
I filed a related issue, #5271, where this problem with `expand_globs` can result in surprising behavior.

**Comment by cirego** [comment]
> Thanks for digging into this so thoroughly! This looks like a bona fide mzbuild bug though. The intent was never that you'd have to run `git add` to get things going. Let me PR a fix at you shortly.

Oh, sounds like your PR might fix #5271 as well then?

File: misc/python/materialize/mzbuild.py [hunk 1]
```diff
@@ -528,6 +530,14 @@ def inputs(self, transitive: bool = False) -> Set[str]:
...
         """
         paths = set(git.expand_globs(self.image.rd.root, f"{self.image.path}/**"))
+        if not paths:
+            # While we could find an `mzbuild.yml` file for this service, git didn't return any
+            # files that matched this service. That likely means that the directory containing the
+            # mzbuild.yml file has not been added to git; otherwise, the mzbuid.yml file, at a
+            # minimum, would be included in the paths list returned by expand_globs.
+            raise errors.BadSpec(
+                f"{self.image.name} mzbuild exists but its files are unknown to git"
+            )
```
Comment (benesch) [inline misc/python/materialize/mzbuild.py:540]: Moving discussion from the other thread here. Basically I think we can never hit this case now, modulo programming bugs? So I think either an assert or an error message that captures that it's an internal mzbuild bug is in order?

Comment (cirego) [reply misc/python/materialize/mzbuild.py:538]: I updated the comment and changed this to an Assertion.

**Review by benesch — APPROVED** [review]
🎉

**Pull request merged**